www/class/config-generate/abstract/host.class.php

```diff
@@ -214,74 +220,298 @@ protected function getHostTemplates(&$host)
         }
     }
 
-    protected function getContacts(&$host)
+    /**
+     * Get contacts list for the configuration file by host
+     *
+     * @param array $host
+     */
+    protected function getContacts(array &$host): void
     {
-        if (!isset($host['contacts_cache'])) {
-            if (is_null($this->stmt_contact)) {
-                $this->stmt_contact = $this->backend_instance->db->prepare("SELECT 
-                    contact_id
-                FROM contact_host_relation
-                WHERE host_host_id = :host_id
-                ");
-            }
-            $this->stmt_contact->bindParam(':host_id', $host['host_id'], PDO::PARAM_INT);
-            $this->stmt_contact->execute();
-            $host['contacts_cache'] = $this->stmt_contact->fetchAll(PDO::FETCH_COLUMN);
+        $contactResult = '';
+        $hostListing = $this->listHostsWithContacts($host);
+        //check if we have Host link to a contact.
+        if (!empty($hostListing)) {
+            $contactResult = implode(',', $this->getInheritanceContact(array_unique($hostListing)));
         }
+        $host['contacts'] = $contactResult;
+    }
 
-        $contact = Contact::getInstance($this->dependencyInjector);
-        $contact_result = '';
-        $contact_result_append = '';
-        foreach ($host['contacts_cache'] as $contact_id) {
-            $tmp = $contact->generateFromContactId($contact_id);
-            if (!is_null($tmp)) {
-                $contact_result .= $contact_result_append . $tmp;
-                $contact_result_append = ',';
+    /**
+     * Get the tree of hosts with contact according to the inheritance notification option
+     *
+     * @param array $host
+     * @return array
+     */
+    public function listHostsWithContacts(array $host): array
```

Comment:
for each host, we don't use cache and browse all the tree ?
if we keep contact cache, we could only parent template ids, and use cache to set contacts.